src/fs.rs

```diff
@@ -551,32 +548,39 @@ impl<IO: Read + Write + Seek, TP, OCC> FileSystem<IO, TP, OCC> {
         Ok(free_cluster_count)
     }
 
-    /// Unmounts the filesystem.
-    ///
-    /// Updates the FS Information Sector if needed.
+    /// Updates the FS Information Sector if needed and unmounts the filesystem.
     ///
     /// # Errors
     ///
     /// `Error::Io` will be returned if the underlying storage object returned an I/O error.
-    pub fn unmount(self) -> Result<(), Error<IO::Error>> {
-        self.unmount_internal()
-    }
+    pub fn unmount(mut self) -> Result<IO, Error<IO::Error>> {
+        self.flush()?;
 
-    fn unmount_internal(&self) -> Result<(), Error<IO::Error>> {
-        self.flush_fs_info()?;
-        self.set_dirty_flag(false)?;
-        Ok(())
+        // SAFETY: The `disk` field is only dropped once.
+        let mut disk = unsafe {
+            let mut fs = ManuallyDrop::new(self);
+            ptr::read(fs.disk.get_mut())
+        };
+        disk.seek(SeekFrom::Start(0))?;
+        Ok(disk)
     }
 
-    fn flush_fs_info(&self) -> Result<(), Error<IO::Error>> {
-        let mut fs_info = self.fs_info.borrow_mut();
+    /// Updates the FS information sector if needed.
+    ///
+    /// # Errors
+    ///
+    /// `Error::Io` will be returned if the underlying storage object returned an I/O error.
+    pub fn flush(&mut self) -> Result<(), Error<IO::Error>> {
+        let fs_info = self.fs_info.get_mut();
         if self.fat_type == FatType::Fat32 && fs_info.dirty {
-            let mut disk = self.disk.borrow_mut();
-            let fs_info_sector_offset = self.offset_from_sector(u32::from(self.bpb.fs_info_sector));
+            let fs_info_sector_offset = self.bpb.bytes_from_sectors(self.bpb.fs_info_sector.into());
+            let disk = self.disk.get_mut();
             disk.seek(SeekFrom::Start(fs_info_sector_offset))?;
-            fs_info.serialize(&mut *disk)?;
+            fs_info.serialize(disk)?;
             fs_info.dirty = false;
         }
+
+        self.set_dirty_flag(false)?;
         Ok(())
```

Comment:
This function should call the underlying `disk.flush();` (I'm currently working on a feature which requires this)